Refactor flatbuffer loader to allow overriding how IValues are parsed.

qihqi opened the pull request:
Test Plan: unittest

Differential Revision: D33720630

zhxchen17 commented on torch/csrc/jit/mobile/flatbuffer_loader.h:
```diff
@@ -50,5 +50,64 @@ TORCH_API mobile::Module load_mobile_module_from_file(
+  FlatbufferLoader();
+
+  typedef IValue (
+      *IValueParser)(FlatbufferLoader*, const mobile::serialization::IValue*);
```
consider `typedef IValue (*IValueParser)(FlatbufferLoader&, const mobile::serialization::IValue&)`. Previously pointer passing was just an implementation detail in loader, but now since we allow overriders here, it's better to make the contract clear that both arguments are non-null values.

qihqi replied:
done.

zhxchen17 commented on torch/csrc/jit/mobile/flatbuffer_loader.h:
```diff
@@ -50,5 +50,64 @@ TORCH_API mobile::Module load_mobile_module_from_file(
+  ClassTypePtr getType(uint32_t pos) {
+    TORCH_CHECK(pos < all_ivalues_.size());
+    return all_types_[pos];
+    // auto iter = all_types_.find(pos);
```
should we remove this?

qihqi replied:
done.

zhxchen17 commented on torch/csrc/jit/mobile/flatbuffer_loader.cpp:
```diff
@@ -196,7 +214,55 @@ std::unique_ptr<mobile::Function> FlatbufferLoader::parseFunction(
+    FlatbufferLoader* loader,
+    const mobile::serialization::IValue* ivalue) {
+  const auto* enum_val = ivalue->val_as_EnumValue();
+  auto enum_type = loader->getOrCreateTypeAnnotations(enum_val->type_name())
```
I'm not 100% sure so just asking: is EnumType even supported on edge?

qihqi replied:
idk... pickle supports it though.

qihqi committed:
Refactor flatbuffer loader to allow overriding how IValues are parsed. (#71661)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/71661

https://docs.google.com/document/d/1OoPKREoqNbOUIcbGzfk8TTIibeTgx3c6Lr3NthF7-PM/edit

Test Plan: unittest

Reviewed By: zhxchen17

Differential Revision: D33720630

fbshipit-source-id: 6f75eb60c41e99e28b086b1adc0a1f7f282159fc
